agent/consul/stream/event.go

```diff
@@ -14,15 +14,23 @@ type Topic fmt.Stringer
 // EventPublisher and returned to Subscribers.
 type Event struct {
 	Topic   Topic
-	Key     string
 	Index   uint64
-	Payload interface{}
+	Payload Payload
+}
+
+type Payload interface {
+	// FilterByKey must return true if the Payload should be included in a subscription
+	// requested with the key and namespace.
+	// Generally this means that the payload matches the key and namespace or
+	// the payload is a special framing event that should be returned to every
+	// subscription.
+	FilterByKey(key, namespace string) bool
```

Comment:
I'm still not sure about having a `namespace` argument like this in OSS rather than following the pattern of passing an `structs.EnterpriseMeta`.

I think practically this isn't really "leaking" much of value, it's more that it's just inconsistent with the way we've abstracted namespaces arguments everywhere else.

The other argument is that we are considering other features which would have wide cross-cutting effects similar to Namespaces and a possible approach right now is to add those additional arguments through the EnterpriseMeta struct we already thread through a lot of code.

All that said, we can relatively easily change this to use EnterpriseMeta later if we do need it for that so I don't think we need to hold up streaming while we decide.

Do you feel this is better this way than using `*structs.EnterpriseMeta` like elsewhere @dnephin ?

Reply:
Bikeshed: it's good this documents the return bool convention clearly, but `FilterByKey` could be read by some as "ShouldWeFilter" which means that `true` would mean the opposite.

I'm OK with it like this since it's documented if people are confused and have to look it up, but I wonder if we could just rename it to make it more intuitive. If it were called `MatchesKey` I think that would be both clearer what the return value means, plus it's more descriptive of the actual logic we expect to be implemented - `FilterBy` implies that the method is actually doing some filtering on the underlying data which it's not for now  (although you maybe hint at doing that instead below...).

Reply:
I tried a few different names here, one of which was `MatchKey`. I think `MatchesKey` sounds better, I'll make that change.

Reply:
> Do you feel this is better this way than using `*structs.EnterpriseMeta` like elsewhere @dnephin ?

Yes, I still think that being explicit about `namespace` here is a better path forward for a few reasons. I think there are a couple ergonomic problems with the current type which we can probably address when we add another field to it. I think the main problem with using EnterpriseMeta here is the dependency graph. `EnterpriseMeta` is in `agent/structs` which has a surprisingly large dependency graph. Using it everywhere means that every package has a transitive dependency on a lot of unrelated things, making it slower to build and work with. If we want to use `EnterpriseMeta` all over the place we should move it to a smaller package like `types`. The code in question here has no dependencies outside the stdlib. It can be re-used and moved around very easily, and test runs are fast (very little time wasted building dependency packages).

I'll also say that there are already a few places where we use `Namespace`. I understand those may have predated EnterpriseMeta, but since they were never cleaned up we aren't really introducing a second way of doing things here.

I think one approach to solving this problem with adjusting EnterpriseMeta would be for the `stream` package to use a `type Key interface` with a method for returning the full ID. If all of our keys implemented that method then `stream` (and other packages) wouldn't need to know anything about EnterpriseMeta. They could all accept opaque keys.

I'll look into that approach, but it may also conflict with #9073